### PR TITLE
ddl: Check the ResolvedTS during the preparation.

### PR DESCRIPTION
**Pull request by HuSharp**
Signed-off-by: husharp <jinhao.hu@pingcap.com>

### What problem does this PR solve?

Issue Number: ref #37197 

Problem Summary:

### What is changed and how it works?
```
Check the ResolvedTS during the preparation to make sure the flashback version is not greater than it.
```
### Check List

Tests 

- [x] Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: ddl/cluster.go [hunk 1]
```diff
@@ -84,6 +85,9 @@ func ValidateFlashbackTS(ctx context.Context, sctx sessionctx.Context, flashBack
 	if oracle.GetTimeFromTS(flashBackTS).After(oracle.GetTimeFromTS(currentTS)) {
 		return errors.Errorf("cannot set flashback timestamp to future time")
 	}
+	if oracle.GetTimeFromTS(flashBackTS).After(expression.GetMinSafeTime(sctx)) {
+		return errors.Errorf("cannot set flashback timestamp to greater than resolved ts from store")
```
Comment (JmPotato) [inline ddl/cluster.go:89]: The term `Resolved TS` could be confusing to the user, maybe we can rephrase it to:

```suggestion
		return errors.Errorf("cannot set flashback timestamp to too close to present")
```

File: executor/recover_test.go [hunk 2]
```diff
@@ -295,6 +297,14 @@ func TestRecoverTableMeetError(t *testing.T) {
 func TestRecoverClusterMeetError(t *testing.T) {
 	store := testkit.CreateMockStore(t)
 	tk := testkit.NewTestKit(t, store)
+	ts, _ := tk.Session().GetStore().GetOracle().GetTimestamp(context.Background(), &oracle.Option{})
+	flashbackTs := oracle.GetTimeFromTS(ts)
+
+	injectSafeTS := oracle.GoTimeToTS(flashbackTs.Add(10 * time.Second))
+	require.NoError(t, failpoint.Enable("tikvclient/injectSafeTS",
+		fmt.Sprintf("return(%v)", injectSafeTS)))
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/expression/injectSafeTS",
```
Comment (JmPotato) [inline executor/recover_test.go:306]: This is unnecessary since it's used to test the `tidb_bounded_staleness` function.

File: executor/recover_test.go [hunk 3]
```diff
@@ -319,21 +329,88 @@ func TestRecoverClusterMeetError(t *testing.T) {
 	// Flashback failed because of ddl history.
 	tk.MustExec("use test;")
 	tk.MustExec("create table t(a int);")
-	tk.MustContainErrMsg(fmt.Sprintf("flashback cluster as of timestamp '%s'", time.Now().Add(0-30*time.Second)), "schema version not same, have done ddl during [flashbackTS, now)")
+	tk.MustContainErrMsg(fmt.Sprintf("flashback cluster as of timestamp '%s'", flashbackTs), "schema version not same, have done ddl during [flashbackTS, now)")
+
+	require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/expression/injectSafeTS"))
```
Comment (JmPotato) [inline executor/recover_test.go:334]: Ditto.

File: executor/recover_test.go (hunk 2 above)
Comment (JmPotato) [reply executor/recover_test.go:306]: It's used to inject the `SafeTS` when the underlying KV store is mocked. Ignore this comment, please.

File: executor/recover_test.go [hunk 4]
```diff
@@ -319,21 +329,88 @@ func TestRecoverClusterMeetError(t *testing.T) {
...
 	tk.MustExec(fmt.Sprintf(safePointSQL, timeBeforeDrop))
 
 	tk.MustContainErrMsg(fmt.Sprintf("flashback cluster as of timestamp '%s'", time.Now().Add(0-30*time.Second)),
 		"not support flash back cluster with TiFlash stores")
+
+	require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/expression/injectSafeTS"))
+	require.NoError(t, failpoint.Disable("tikvclient/injectSafeTS"))
+}
+
+func TestRecoverClusterWithSafeTs(t *testing.T) {
```
Comment (JmPotato) [inline executor/recover_test.go:361]:
```suggestion
func TestFlashbackWithSafeTs(t *testing.T) {
```

**Review by JmPotato — COMMENTED** [review]
The reset LGTM. PTAL @Defined2014 @bb7133

**Review by JmPotato — APPROVED** [review]

**Comment by Defined2014** [comment]
LGTM, Need fix some UTs, like `TestFlashbackCloseAndResetPDSchedule`  and `TestCancelFlashbackCluster`

**Comment by HuSharp** [comment]
Fixed tests that include `flashback cluster as of timestamp` in the `cluster_test.go` and `recover_test.go` files. PTAL @Defined2014 @JmPotato @bb7133

**Review by Defined2014 — APPROVED** [review]
